src/tscpp/api/TransformationPlugin.cc

```diff
@@ -273,31 +285,44 @@ TransformationPlugin::~TransformationPlugin()
   delete state_;
 }
 
-TSCont
+void
 TransformationPlugin::pause()
 {
-  if (state_->input_complete_dispatched_) {
+  if (state_->paused_) {
+    LOG_ERROR("Can not pause transformation, already paused  TransformationPlugin=%p (vconn)contp=%p tshttptxn=%p", this,
+              state_->vconn_, state_->txn_);
+  } else if (state_->input_complete_dispatched_) {
     LOG_ERROR("Can not pause transformation (transformation completed) TransformationPlugin=%p (vconn)contp=%p tshttptxn=%p", this,
               state_->vconn_, state_->txn_);
-    return nullptr;
   } else {
-    state_->paused_      = true;
-    state_->resume_cont_ = TSContCreate(&resumeCallback, TSContMutexGet(reinterpret_cast<TSCont>(state_->txn_)));
-    TSContDataSet(state_->resume_cont_, static_cast<void *>(state_));
-    return state_->resume_cont_;
+    state_->paused_ = true;
+    if (!static_cast<bool>(state_->resume_cont_)) {
+      state_->resume_cont_ = TransformationPluginContinuation(TSContMutexGet(reinterpret_cast<TSCont>(state_->txn_)));
+    }
   }
 }
 
+bool
+TransformationPlugin::isPaused() const
+{
+  return state_->paused_;
+}
+
+Continuation &
+TransformationPlugin::resumeCont()
+{
+  TSReleaseAssert(state_->paused_);
+
+  return state_->resume_cont_;
+}
+
 int
-TransformationPlugin::resumeCallback(TSCont cont, TSEvent event, void *edata)
+TransformationPluginContinuation::_run(TSEvent event, void *edata)
 {
-  auto state = static_cast<TransformationPluginState *>(TSContDataGet(cont));
-  if (state) {
-    state->paused_ = false;
-    handleTransformationPluginRead(state->vconn_, state);
-  }
+  auto state     = TS_FIELD_TO_CLASS_PTR(TransformationPluginState, resume_cont_, this);
```

Comment:
Seems like a bit of a large change just to avoid a `TSContDataGet`.

Reply:
In an object-oriented system, you call the methods of objects, e.g. obj.foo() .  If the system is also event-driven, foo() may need to complete it's actions after one or a series of events happen.  So we need some approach to handling this.  This PR is an example of my preferred approach so far.  I could have instead made TransformationPluginContinuation a  base class of TransformationPluginState, and downcasted.  In the generated machine code, downcasting and using the TS_FIELD_TO_CLASS_PTR macro would be the same.  Ultimately it's largely an aesthetic choice.  The line above would become the downcast:

auto state = static_cast<TransformationPluginState *>(this);

I think it adds confusion to use base classes for something other than a clear "is-a" relationship.

Reply:
@ywkaras and I discussed this at length. My view is that subclassing `Continuation` and putting the back pointer there would be the best approach. That's a standard C++ style approach to this problem.

Reply:
Party pooper.